### PR TITLE
[Merged by Bors] - Feat: port Order.Filter.Bases

### PR DESCRIPTION
**Pull request by urkud**
---

- [x] depends on: #1788

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#1788
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1788~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Order/Filter/Bases.lean [hunk 1]
```diff
@@ -8,120 +8,130 @@ Authors: Yury Kudryashov, Johannes Hölzl, Mario Carneiro, Patrick Massot
...
-the proposition `h : filter.is_basis p s` makes sure the range of `s` bounded by `p`
+the proposition `h : Filter.IsBasis p s` makes sure the range of `s` bounded by `p`
 (ie. `s '' set_of p`) defines a filter basis `h.filter_basis`.
 
-If one already has a filter `l` on `α`, `filter.has_basis l p s` (where `p : ι → Prop`
+If one already has a filter `l` on `α`, `Filter.HasBasis l p s` (where `p : ι → Prop`
 and `s : ι → set α` as above) means that a set belongs to `l` if and
-only if it contains some `s i` with `p i`. It implies `h : filter.is_basis p s`, and
+only if it contains some `s i` with `p i`. It implies `h : Filter.IsBasis p s`, and
 `l = h.filter_basis.filter`. The point of this definition is that checking statements
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:32]:
```suggestion
`l = h.filterBasis.filter`. The point of this definition is that checking statements
```

File: Mathlib/Order/Filter/Bases.lean [hunk 2]
```diff
@@ -0,0 +1,1214 @@
...
+the proposition `h : Filter.IsBasis p s` makes sure the range of `s` bounded by `p`
+(ie. `s '' set_of p`) defines a filter basis `h.filter_basis`.
+
+If one already has a filter `l` on `α`, `Filter.HasBasis l p s` (where `p : ι → Prop`
+and `s : ι → set α` as above) means that a set belongs to `l` if and
+only if it contains some `s i` with `p i`. It implies `h : Filter.IsBasis p s`, and
+`l = h.filter_basis.filter`. The point of this definition is that checking statements
+involving elements of `l` often reduces to checking them on the basis elements.
+
+We define a function `has_basis.index (h : Filter.HasBasis l p s) (t) (ht : t ∈ l)` that returns
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:35]:
```suggestion
We define a function `HasBasis.index (h : Filter.HasBasis l p s) (t) (ht : t ∈ l)` that returns
```

File: Mathlib/Order/Filter/Bases.lean [hunk 3]
```diff
@@ -0,0 +1,1214 @@
...
+  nonempty : ∃ i, p i
+  /-- `s` is directed downwards on `i` such that `p i`. -/
+  inter : ∀ {i j}, p i → p j → ∃ k, p k ∧ s k ⊆ s i ∩ s j
+#align filter.is_basis Filter.IsBasis
+
+namespace Filter
+
+namespace IsBasis
+
+/-- Constructs a filter basis from an indexed family of sets satisfying `is_basis`. -/
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:142]:
```suggestion
/-- Constructs a filter basis from an indexed family of sets satisfying `IsBasis`. -/
```

File: Mathlib/Order/Filter/Bases.lean [hunk 4]
```diff
@@ -0,0 +1,1214 @@
...
+
+end FilterBasis
+
+namespace Filter
+
+namespace IsBasis
+
+variable {p : ι → Prop} {s : ι → Set α}
+
+/-- Constructs a filter from an indexed family of sets satisfying `is_basis`. -/
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:214]:
```suggestion
/-- Constructs a filter from an indexed family of sets satisfying `IsBasis`. -/
```

File: Mathlib/Order/Filter/Bases.lean [hunk 5]
```diff
@@ -0,0 +1,1214 @@
...
+    exact ⟨⋂₀ (a ∪ b), mem_image_of_mem _ ⟨fina.union finb, union_subset suba subb⟩,
+        (interₛ_union _ _).subset⟩
+#align filter.filter_basis.of_sets Filter.FilterBasis.ofSets
+
+lemma FilterBasis.ofSets_sets (s : Set (Set α)) :
+    (FilterBasis.ofSets s).sets = interₛ '' { t | Set.Finite t ∧ t ⊆ s } :=
+  rfl
+
+-- porting note: use `∃ i, p i ∧ _` instead of `∃ i (hi : p i), _`.
+/-- Definition of `has_basis` unfolded with implicit set argument. -/
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:264]:
```suggestion
/-- Definition of `HasBasis` unfolded with implicit set argument. -/
```

File: Mathlib/Order/Filter/Bases.lean [hunk 6]
```diff
@@ -0,0 +1,1214 @@
...
+  simpa only [infᵢ_true] using h.eq_binfᵢ
+#align filter.has_basis.eq_infi Filter.HasBasis.eq_infᵢ
+
+theorem hasBasis_infᵢ_principal {s : ι → Set α} (h : Directed (· ≥ ·) s) [Nonempty ι] :
+    (⨅ i, 𝓟 (s i)).HasBasis (fun _ => True) s :=
+  ⟨fun t => by
+    simpa only [true_and] using mem_infᵢ_of_directed (h.mono_comp monotone_principal.dual) t⟩
+#align filter.has_basis_infi_principal Filter.hasBasis_infᵢ_principal
+
+/-- If `s : ι → set α` is an indexed family of sets, then finite intersections of `s i` form a basis
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:768]:
```suggestion
/-- If `s : ι → Set α` is an indexed family of sets, then finite intersections of `s i` form a basis
```

File: Mathlib/Order/Filter/Bases.lean [hunk 7]
```diff
@@ -0,0 +1,1214 @@
...
+#align filter.has_basis.bInter_mem Filter.HasBasis.bInter_mem
+
+theorem HasBasis.interₛ_sets (h : HasBasis l p s) : ⋂₀ l.sets = ⋂ (i) (_hi : p i), s i := by
+  rw [interₛ_eq_binterᵢ]
+  exact h.bInter_mem monotone_id
+#align filter.has_basis.sInter_sets Filter.HasBasis.interₛ_sets
+
+variable {ι'' : Type _} [Preorder ι''] (l) (s'' : ι'' → Set α)
+
+/-- `is_antitone_basis s` means the image of `s` is a filter basis such that `s` is decreasing. -/
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:835]:
```suggestion
/-- `IsAntitoneBasis s` means the image of `s` is a filter basis such that `s` is decreasing. -/
```

File: Mathlib/Order/Filter/Bases.lean [hunk 8]
```diff
@@ -0,0 +1,1214 @@
...
+structure IsAntitoneBasis extends IsBasis (fun _ => True) s'' : Prop where
+  /-- The sequence of sets is antitone. -/
+  protected antitone : Antitone s''
+#align filter.is_antitone_basis Filter.IsAntitoneBasis
+
+/-- We say that a filter `l` has an antitone basis `s : ι → set α`, if `t ∈ l` if and only if `t`
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:841]:
```suggestion
/-- We say that a filter `l` has an antitone basis `s : ι → Set α`, if `t ∈ l` if and only if `t`
```

File: Mathlib/Order/Filter/Bases.lean [hunk 9]
```diff
@@ -0,0 +1,1214 @@
...
+
+end Filter
+
+end sort
+
+namespace Filter
+
+variable {α β γ ι : Type _} {ι' : Sort _}
+
+/-- `is_countably_generated f` means `f = generate s` for some countable `s`. -/
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:979]:
```suggestion
/-- `IsCountablyGenerated f` means `f = generate s` for some countable `s`. -/
```

File: Mathlib/Order/Filter/Bases.lean [hunk 10]
```diff
@@ -0,0 +1,1214 @@
...
+class IsCountablyGenerated (f : Filter α) : Prop where
+  /-- There exists a countable set that generates the filter. -/
+  out : ∃ s : Set (Set α), s.Countable ∧ f = generate s
+#align filter.is_countably_generated Filter.IsCountablyGenerated
+
+/-- `is_countable_basis p s` means the image of `s` bounded by `p` is a countable filter basis. -/
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:985]:
```suggestion
/-- `IsCountableBasis p s` means the image of `s` bounded by `p` is a countable filter basis. -/
```

File: Mathlib/Order/Filter/Bases.lean [hunk 11]
```diff
@@ -0,0 +1,1214 @@
...
+
+/-- A countable filter basis `B` on a type `α` is a nonempty countable collection of sets of `α`
+such that the intersection of two elements of this collection contains some element
+of the collection. -/
+structure CountableFilterBasis (α : Type _) extends FilterBasis α where
+  /-- The set of sets of the filter basis is countable. -/
+  countable : sets.Countable
+#align filter.countable_filter_basis Filter.CountableFilterBasis
+
+-- For illustration purposes, the countable filter basis defining (at_top : filter ℕ)
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:1008]:
```suggestion
-- For illustration purposes, the countable filter basis defining `(AtTop : Filter ℕ)`
```

File: Mathlib/Order/Filter/Bases.lean [hunk 12]
```diff
@@ -0,0 +1,1214 @@
...
+  hs.toHasBasis.mem_of_mem trivial
+#align filter.has_antitone_basis.mem Filter.HasAntitoneBasis.mem
+
+theorem HasAntitoneBasis.hasBasis_ge [Preorder ι] [IsDirected ι (· ≤ ·)] {l : Filter α}
+    {s : ι → Set α} (hs : l.HasAntitoneBasis s) (i : ι) : l.HasBasis (fun j => i ≤ j) s :=
+  hs.1.to_hasBasis (fun j _ => (exists_ge_ge i j).imp fun _k hk => ⟨hk.1, hs.2 hk.2⟩) fun j _ =>
+    ⟨j, trivial, Subset.rfl⟩
+#align filter.has_antitone_basis.has_basis_ge Filter.HasAntitoneBasis.hasBasis_ge
+
+/-- If `f` is countably generated and `f.has_basis p s`, then `f` admits a decreasing basis
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Bases.lean:1069]:
```suggestion
/-- If `f` is countably generated and `f.HasBasis p s`, then `f` admits a decreasing basis
```

**Review by jcommelin — COMMENTED** [review]
bors d+

**Comment by urkud** [comment]
bors merge

**Pull request closed without merging**